includes/class-llms-loader.php

```diff
@@ -230,11 +230,15 @@ public function autoload( $class ) {
 	 * @since 5.6.0 Include `LLMS_Prevent_Concurrent_Logins`.
 	 * @since 6.0.0 Included `LLMS_Block_Library`, `LLMS_Controller_Awards`, and `LLMS_Engagement_Handler`.
 	 *              Removed loading of class files that don't instantiate their class in favor of autoloading.
+	 * @since [version] Instantiated `LLMS_Shortcodes` before `LLMS_Controller_Orders`.
 	 *
 	 * @return void
 	 */
 	public function includes() {
 
+		// Instantiate LLMS_Shortcodes before LLMS_Controller_Orders.
+		new LLMS_Shortcodes();
```

Comment:
I know why this is done this way but instantiating a class in a method that only includes files seems like a break in an established pattern. I know the autoloader loads it and it performs the same function but I don't like it. I'd prefer to instantiate elsewhere instead of a method with an explicit purpose of "including files"

Reply:
I don't like it either for the same reasons. However, the purpose of this method changed in 6.0.0 from "include a bunch of classes used by almost every page and instantiate some of those classes so they can hook into the system" to "include only the classes that instantiate themselves so they can hook into the system". Loading the class file is not the goal. It's just a way to get to the goal of initializing the plug-in.

I'd like to rename this method, but it's public. Maybe we should deprecate this method and move its code to a private/protected "bootstrap" or "init" function/method? It belongs in the core `LifterLMS` class and should be called by the `init()` method.

I didn't want to go too deep on this because the timing of dependencies during startup can be tricky and time-consuming. So I "settled" for a method that, while doing essentially the same thing, isn't quite named accurately.

I would like to make it a rule that we never instantiate an object in a class definition file. That's a rule I've lived by forever, but I don't remember where I learned it. However, I need to think hard about how this change might affect 3rd parties using these classes.

Reply:
From [PSR-1: Basic Coding Standard](https://www.php-fig.org/psr/psr-1/):
> Files SHOULD either declare symbols (classes, functions, constants, etc.) or cause side-effects (e.g. generate output, change .ini settings, etc.) but SHOULD NOT do both.

Reply:
> I would like to make it a rule that we never instantiate an object in a class definition file. That's a rule I've lived by forever, but I don't remember where I learned it. However, I need to think hard about how this change might affect 3rd parties using these classes.

This rule introduces a new standard that breaks a substantial (I'd say *most*) files in our codebases. So we're not going to introduce this as a rule. PSR-1 is great I guess (I don't know much about any of the established standards). This pattern breaks it. Okay. But we're not going to just change a bunch of stuff and create a new rule because of this.


> I'd like to rename this method, but it's public. Maybe we should deprecate this method and move its code to a private/protected "bootstrap" or "init" function/method? It belongs in the core LifterLMS class and should be called by the init() method.

The main `LifterLMS` class already has this bootstrapping for a bunch of classes and so on. And shortcodes are *already* being bootstrapped there.

Reply:
If the shortcode class initializes itself (willfully breaking the PSR-1 standard but following our pre-established patterns) the require can take place in the `includes()` method without me feeling weird about it and I think we're probably good to go on these changes.

Reply:
OK.

Reply:
> > I would like to make it a rule that we never instantiate an object in a class definition file. That's a rule I've lived by forever, but I don't remember where I learned it. However, I need to think hard about how this change might affect 3rd parties using these classes.
> 
> This rule introduces a new standard that breaks a substantial (I'd say _most_) files in our codebases. So we're not going to introduce this as a rule. PSR-1 is great I guess (I don't know much about any of the established standards). This pattern breaks it. Okay. But we're not going to just change a bunch of stuff and create a new rule because of this.

When I said, "I would like to make it a rule ...", I didn't mean that I think the rule should start right now. I should have created a feature request issue instead of suggesting it in a PR. It would be a waste of time to change the pattern just because someone "likes" it a certain way, so I'll include justifications. 

Changing how we bootstrap the plugin from the current standard to the PSR-1 standard is a relatively simple change.
1. Remove the `return new CLASS();` from 44 files.
2. Replace `LLMS_Loader::includes()` that includes the files with `LLMS::init_classes()` (or something like that) that instantiates the classes. This method should be called during LifterLMS bootstrapping, possibly by `LLMS::init()`.